src/AST.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE TupleSections   #-}
+{-# LANGUAGE ConstrainedClassMethods   #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+{-# LANGUAGE FlexibleInstances   #-}
+module AST
+  ( ASTTable(..)
+  , ASTBuildError(..)
+  , buildAST
+  , buildASTTable
+  , buildASTTable'
+  , getEntrypoint
+  )
+where
+
+import           Grammar                        ( parse
+                                                , AST(aimports, apath)
+                                                , ImportDecl(ipath)
+                                                , Path
+                                                )
+import qualified Data.Map                      as M
+import           Control.Monad                  ( liftM2 )
+import           Data.Either                    ( fromRight )
+import           Control.Exception              ( IOException
+                                                , try
+                                                )
+import           Data.Either.Combinators        ( mapLeft )
+import           System.FilePath.Posix          ( splitFileName )
+
+type ASTTable = M.Map FilePath AST
+
+data ASTBuildError = ImportNotFound FilePath (Maybe AST)
+                   | LexicalError FilePath String
+                   | EntrypointNotFound FilePath
+                   deriving(Eq, Show)
+
+
+buildASTTable :: FilePath -> IO (Either ASTBuildError ASTTable)
+buildASTTable path =
+  let rootPath = computeRootPath path
+  in  buildASTTable' readFile Nothing rootPath path
+
+-- TODO: Remove dupplicate with Resolver
+computeRootPath :: FilePath -> FilePath
+computeRootPath = fst . splitFileName
+
+buildASTTable'
+  :: (FilePath -> IO String)
+  -> Maybe AST
+  -> FilePath
+  -> FilePath
+  -> IO (Either ASTBuildError ASTTable)
+buildASTTable' rf parent rootPath entrypoint = do
+  s <- try $ rf entrypoint :: IO (Either IOException String)
+  let source = either (const $ Left $ ImportNotFound entrypoint parent) Right s
+      ast = source >>= mapLeft (LexicalError entrypoint) . buildAST entrypoint
+      importPaths = importPathsFromAST rootPath ast
+      generatedTable = uncurry M.singleton . (entrypoint, ) <$> ast
+      nextParent = either (const Nothing) Just ast
+  childTables <- mapM (buildASTTable' rf nextParent rootPath) importPaths
+  return $ foldl (liftM2 $ flip M.union) generatedTable childTables
```

Comment:
For me almost everything else is ~readable, but this block is a little hard to comprehend. It kinda has to make sense, relative to the goals of it, but I find the content here pretty densely packed.

Reply:
I get you, I'll see how I can refactor that to make it more comprehensive then !

One thought I had was to move line 56 in buildAST. Basically I just `Either String AST` coming from buildAST into an `Either ASTError AST` that has more context. So line 56 can just become:
`ast = source >>= buildAST entrypoint`

That's still one of my struggle. I feel like the code I write in Haskell is often not very readable. But maybe it's just my inability to parse it correctly yet.

Reply:
Oh wait, I did it already haha. Well, but definitely it's dense. I could potentially add types to make it clearer, or split it in simpler functions that have names for clarity.